leap_c/trainer.py

```diff
@@ -399,37 +406,86 @@ def _ckpt_path(self, name: str, suffix: str, basedir: Path | None = None) -> Pat
         if basedir is None:
             basedir = self.output_path
 
+        if basedir is None:
+            raise ValueError("Output path is not set. Either set it in the init or pass it to the function.")
+
         (basedir / "ckpts").mkdir(exist_ok=True)
 
         if self.cfg.val.ckpt_modus == "best":
             return basedir / "ckpts" / f"best_{name}.{suffix}"
         elif self.cfg.val.ckpt_modus == "last":
             return basedir / "ckpts" / f"last_{name}.{suffix}"
 
-        return basedir / "ckpts" / f"{self.step}_{name}.{suffix}"
+        return basedir / "ckpts" / f"{self.state.step}_{name}.{suffix}"
+
+    def save(self, path: Path | None = None) -> None:
+        """Save the trainer state in a checkpoint folder.
 
-    def save(self) -> None:
-        """Save the trainer state in a checkpoint folder."""
+        If the path is None, the checkpoint is saved in the output path of the trainer.
+        The state_dict is split into different parts. For example if the trainer has
+        as submodule "pi" and "q", the state_dict is saved separately as "pi.ckpt" and
+        "q.ckpt". Additionally, the optimizers are saved as "optimizers.ckpt" and the
+        trainer state is saved as "trainer_state.ckpt".
+
+        Args:
+            path: The folder where to save the checkpoint.
+        """
+
+        # split the state_dict into seperate parts
+        split_state_dicts = defaultdict(dict)
+        for name, param in self.state_dict().items():
+            if "." in name:
+                group_name = name.split(".")[0]
+                sub_name = ".".join(name.split(".")[1:])
+                split_state_dicts[group_name][sub_name] = param
+            else:
+                split_state_dicts[name] = param  # type: ignore
 
-        torch.save(self.state_dict(), self._ckpt_path("model", "pth"))
-        torch.save(self.state, self._ckpt_path("trainer", "pkl"))
+        # save the state dicts
+        for name, state_dict in split_state_dicts.items():
+            torch.save(state_dict, self._ckpt_path(name, "ckpt", path))
+
+        torch.save(self.state, self._ckpt_path("trainer_state", "ckpt", path))
 
         if self.optimizers:
             state_dict = {
                 f"optimizer_{i}": opt.state_dict()
                 for i, opt in enumerate(self.optimizers)
             }
-            torch.save(state_dict, self._ckpt_path("optimizers", "pth"))
+            torch.save(state_dict, self._ckpt_path("optimizers", "ckpt", path))
 
-    def load(self, path: str | Path) -> None:
-        """Loads the state of a trainer from the output_path."""
-        # TODO (Jasper): Think about partial loading...
-        basedir = Path(path) 
+    def load(self, path: Path) -> None:
```

Comment:
Why not allow str, when we convert it to path in line 463 anyways? Also we allow str in the init for output path? (Im not for or against allowing string, but I think it should be kept consistent)